rollout/restart.go

```diff
@@ -0,0 +1,163 @@
+package rollout
+
+import (
+	"sort"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/argoproj/argo-rollouts/utils/replicaset"
+)
+
+const (
+	// restartPodCheckTime prevents the Rollout from not making any progress with restarting Pods. When pods can be restarted
+	// faster than the old pods can be scaled down, the parent's ReplicaSet's availableReplicas does not change. A rollout
+	// uses changes to the availableReplicas of the ReplicaSet to detect when the controller should requeue and continue
+	// deleting pods. In this situation, the rollout does not renqueue and wont make any more progress restarting pods until
+	// the resync period passes or another change is made to the Rollout. The controller requeue Rollouts with deleted
+	// Pods every 30 seconds to make sure the rollout is not stuck.
+	restartPodCheckTime = 30 * time.Second
+)
+
+// RolloutPodRestarter describes the components needed for the controller to restart all the pods of
+// a rollout.
+type RolloutPodRestarter struct {
+	client       kubernetes.Interface
+	resyncPeriod time.Duration
+	enqueueAfter func(obj interface{}, duration time.Duration)
+}
+
+// checkEnqueueRollout enqueues a Rollout if the Rollout's restartedAt is within the next resync
+func (p RolloutPodRestarter) checkEnqueueRollout(roCtx rolloutContext) {
+	r := roCtx.Rollout()
+	logCtx := roCtx.Log().WithField("Reconciler", "PodRestarter")
+	now := nowFn()
+	if r.Spec.RestartAt == nil || now.After(r.Spec.RestartAt.Time) {
+		return
+	}
+	nextResync := now.Add(p.resyncPeriod)
+	// Only enqueue if the Restart time is before the next sync period
+	if nextResync.After(r.Spec.RestartAt.Time) {
+		timeRemaining := r.Spec.RestartAt.Sub(now)
+		logCtx.Infof("Enqueueing Rollout in %s seconds for restart", timeRemaining.String())
+		p.enqueueAfter(r, timeRemaining)
+	}
+}
+
+func (p *RolloutPodRestarter) Reconcile(roCtx rolloutContext) error {
+	rollout := roCtx.Rollout()
+	logCtx := roCtx.Log().WithField("Reconciler", "PodRestarter")
+	p.checkEnqueueRollout(roCtx)
+	if !replicaset.NeedsRestart(rollout) {
+		return nil
+	}
+	logCtx.Info("Reconcile pod restarts")
+	s := NewSortReplicaSetsByPriority(roCtx)
+	for _, rs := range s.allRSs {
+		if rs.Status.AvailableReplicas != *rs.Spec.Replicas {
+			logCtx.WithField("ReplicaSet", rs.Name).Info("cannot restart pods as not all ReplicasSets are fully available")
+			return nil
+		}
+	}
+	sort.Sort(s)
+	for _, rs := range s.allRSs {
+		reconciledReplicaSet, err := p.reconcilePodsInReplicaSet(roCtx, rs)
+		if err != nil {
+			return err
+		}
+		if reconciledReplicaSet {
+			return nil
+		}
+	}
+	logCtx.Info("all pods have been restarted and setting restartedAt status")
+	roCtx.SetRestartedAt()
+	return nil
+}
+
+func (p RolloutPodRestarter) reconcilePodsInReplicaSet(roCtx rolloutContext, rs *appsv1.ReplicaSet) (bool, error) {
```

Comment:
Could you add comment what the purpose of this method is, and the meaning of a true/false return value.

I think we should rename this to indicate what it's doing since `reconcilePodsInReplicaSet` is nondescript (e.g. `restartReplicaSetPod`). I actually initially thought the `reconciledReplicaSet` boolean value meant that all pods are restarted properly, but it actually means the opposite.